awscli/customizations/eks/update_kubeconfig.py

```diff
@@ -229,14 +230,24 @@ def _expand_path(self, path):
 
 
 class EKSClient(object):
-    def __init__(self, session, cluster_name, role_arn, parsed_globals=None):
+    # A set that contains optional flag keys here. 
+    # If you need to specify a default value other than `None` for the parsed_args,
+    # you need to provide a value for the 'default' key in ARG_TABLE. 
+    _OPTIONAL_ATTRS = {
+        'role_arn',
+    }
+
+    def __init__(self, session, *, cluster_name, parsed_globals=None, **parsed_args):
```

Comment:
This is honestly a prime place for using `dataclasses` - providing optional parameters and defaults. Here are the docs for Python 3.7 `dataclasses` - https://docs.python.org/3.7/library/dataclasses.html

I don't see them being used anywhere else in the AWS CLI v2 package, though, but they should be supported?

Reply:
Thanks for this pointer! Glad I've learned something new. So I've looked into `dataclasses` and apparently it has some limitations, in order of importance: 
1. is equivalent to setting all fields as named arguments
  - this requires setting *each* field explicitly, e.g. `EKSClient(..., role_arn=parsed_args.role_arn)`
  - still changes the init method signature as data fields are added
    - this leads to updating the code at multiple places, therefore not solving the problems targeted initially
  - it cannot take additional arguments and discard them later like in `EKSClient(..., **vars(parsed_args)` (parsed_args contains extra arguments like `kubeconfig` that will error)
2. requires explicit init method if we want to hide the attribute
  - we could do an explicit `self._session = session` with `InitVar`s
    - this seems to defeat the purpose of using `dataclasses`
  - we could name the data field `_session` and call with `EKSClient(_session=xyz)` but this seems to defeat the purpose of hiding it

In contrast, the limitations of my current approach: 
1. explicitly setting non-public attributes with `setattr()` seems to be bad practice
2. passing extra kwargs that `EKSClient` doesn't need
3. usage of this dict is not obvious

Happy to learn anything I've overlooked.

Reply:
@RedbackThomson Hi!
What is the cause that is currently stopping this? 👀

Reply:
@kahirokunn My comment was purely a code style suggestion. It certainly should not block the merge of this PR

Reply:
`dataclasses` is definitely an option as v1 now only supports Python 3.7+. I think I would have opted with this approach if we were not concerned with the bloat of initializer arguments and plumbing parsed arguments through.

One question that I have is there a particular reason we just accept the `parsed_args` as is (without expanding them as keyword arguments) similar to how we accept `parsed_globals`? I sort of like this approach for a couple of reasons:
1. It makes even less code we need to the `EKSClient` class every time a new argument is added; we would not even need to add to the `_OPTIONAL_ATTRS` class set.
2. It is very clear what arguments are actually accepted to the initializer of the `EKSClient` class. In reading the class as it is now, readers have to be aware of the `_OPTIONAL_ATTRS` class property in trying to determine what keyword arguments are respected by the client.

Reply:
Thank you, makes a ton of sense. I think I spent too much effort trying to work around the existing usage of `cluster_name` and `role_arn`. The `ARG_TABLE` is and should be kept as the single source of truth for the args.